src/EditorFeatures/Core.Wpf/InlineDiagnostics/InlineDiagnosticsTaggerProvider.cs

```diff
@@ -0,0 +1,101 @@
+using System.Collections.Generic;
+using System.ComponentModel.Composition;
+using System.Diagnostics;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Editor.Implementation.Diagnostics;
+using Microsoft.CodeAnalysis.Editor.Implementation.EditAndContinue;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Shared.TestHooks;
+using Microsoft.VisualStudio.Text.Adornments;
+using Microsoft.VisualStudio.Text.Classification;
+using Microsoft.VisualStudio.Text.Tagging;
+using Microsoft.VisualStudio.Utilities;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Editor.InlineDiagnostics
+{
+    [Export(typeof(ITaggerProvider))]
+    [ContentType(ContentTypeNames.RoslynContentType)]
+    [TagType(typeof(InlineDiagnosticsTag))]
+    internal class InlineDiagnosticsTaggerProvider : AbstractDiagnosticsAdornmentTaggerProvider<InlineDiagnosticsTag>
+    {
+        private readonly IEditorFormatMap _editorFormatMap;
+        protected sealed override IEnumerable<PerLanguageOption2<bool>> PerLanguageOptions => SpecializedCollections.SingletonEnumerable(InlineDiagnosticsOptions.EnableInlineDiagnostics);
+
+        protected internal override bool IsEnabled => true;
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public InlineDiagnosticsTaggerProvider(
+            IThreadingContext threadingContext,
+            IEditorFormatMapService editorFormatMapService,
+            IDiagnosticService diagnosticService,
+            IAsynchronousOperationListenerProvider listenerProvider)
+            : base(threadingContext, diagnosticService, listenerProvider)
+        {
+            _editorFormatMap = editorFormatMapService.GetEditorFormatMap("text");
+        }
+
+        protected internal override bool IncludeDiagnostic(DiagnosticData diagnostic)
+        {
+            return
+                diagnostic.Severity is DiagnosticSeverity.Warning or DiagnosticSeverity.Error &&
+                !string.IsNullOrWhiteSpace(diagnostic.Message) &&
+                !diagnostic.IsSuppressed;
+        }
+
+        /// <summary>
+        /// Creates the InlineErrorTag with the error distinction
+        /// </summary>
+        protected override InlineDiagnosticsTag? CreateTag(Workspace workspace, DiagnosticData diagnostic)
+        {
+            Debug.Assert(!string.IsNullOrWhiteSpace(diagnostic.Message));
+            var errorType = GetErrorTypeFromDiagnostic(diagnostic);
+            if (errorType is null)
+            {
+                return null;
+            }
+
+            if (diagnostic.DocumentId is null)
+            {
+                return null;
+            }
+
+            var document = workspace.CurrentSolution.GetRequiredDocument(diagnostic.DocumentId);
+            if (document is null)
```

Comment:
can't be null, you called GetRequiredDocument above.  either this is always non null and you don't nee dthe check.  or the doc could not be found, in which case you need GetDocument instead.

Reply:
changed to getDocument